### PR TITLE
Downloading route content (Github, zip)

### PR DESCRIPTION
**Pull request by sweiland-openrails**
[Trello card](https://trello.com/c/1YXKK7N5/164-publishing-and-downloading-content)

I've created within the Main menu a separate window "Download Content". So that it can be enlarged when more download options would to be added (trains, consists, etc.). Just like the Resume/Replay.

image

Wishes for the second phase:

- [X] add uninstall button
- [X] add info button, starts webpage
- [X] replace hardcoded GitHub URL's by downloading a file with it from GitHub Open Rails
- [X] add check for diskspace and tell how much diskspace is used by the download. "Download" might not be the right verb, "Install" to be used?
- [X] progress bar, there is no such thing as a callback for every MB downloaded from the repository Clone. I might be able to get the total amount Mb to be downloaded versus the already downloaded MB in a progress bar. And have the download process running in a different thread. 
- [X] load the just downloaded route in the menu as the to be started route. Or even start the route.
- [X] add a update button and beforehand show what's get updated.

### CODE REVIEW

File: Source/Documentation/Manual/start.rst [hunk 1]
```diff
@@ -16,7 +16,24 @@ installation profile.
...
+.. image:: images/download-content-form.png
+
+Select the route to be installed. This list will be enlarged by the OR team when suitable routes become available.
+However it is still possible to install routes manually as described in the next chapter.
+
+* ``Install path`` and ``Browse``: Fill in or Browse to the path where the route is to be installed. Each route will be installed in it's own subdirectory.
+* ``Info``: Gives more information of the selected route. The default web browser will be started with the information page.
+* ``Install``: Installs the selected route. A popup mentions the directory where the route is to be installed before starting the actual install. Installation profile gets added also.
+* ``Start``: Some routes have default start parameters and can be started via this button with these parameters.
+* ``Delete``: Delete route with this button. Take care: whole directory including added and changed files gets deleted.
```
Comment (cjakeman) [inline Source/Documentation/Manual/start.rst:35]:
```suggestion
* ``Delete``: Delete route with this button. Take care: whole directory including any added and changed files gets deleted.
```

File: Source/Documentation/Manual/start.rst [hunk 2]
```diff
@@ -16,7 +16,24 @@ installation profile.
...
+* ``Update``: GitHub based routes can get updated after the install by means of this button. Check for updates is done when clicking this button. Or at the end of the information webpage. Also GitHub has a watch possibility for getting changes on the repository being mailed.
```
Comment (cjakeman) [inline Source/Documentation/Manual/start.rst:36]:
```suggestion
* ``Update``: GitHub-based routes can get updated after the install by means of this button. A check for updates is made when clicking this button and updates may be reported at the end of the information webpage. Also GitHub has a watch feature which notifies you of changes by email.
```

**Comment by sweiland-openrails** [comment]
hi Chris, to my opinion this enhancement is ready for a final review and merge. Do you agree?

**Comment by cjakeman** [comment]
Hi Siebren,

> hi Chris, to my opinion this enhancement is ready for a final review and merge. Do you agree?

It all seems to be working well. I spoke with James about it on Saturday and we want to move Options > Content into a tab on your new form before it gets released. All the content would then be accessible via the "Content" button.

I'll post some more at the weekend to see what your thoughts are. James and I are happy to do some of the work if necessary to get this released.

Best wishes,

Chris

File: Source/Menu/DownloadContentForm.cs [hunk 3]
```diff
@@ -0,0 +1,1270 @@
...
+            dataGridViewDownloadContent.CurrentRow.Cells[1].Value = dateTimeNow.ToString(CultureInfo.CurrentCulture.DateTimeFormat);
+            route.DateInstalled = dateTimeNow;
+
+            route.DirectoryInstalledIn = installPathRoute;
+
+            Settings.Folders.Save();
+            Settings.Content.Save();
+
+            MainForm mainForm = ((MainForm)Owner);
+            mainForm.LoadFolderListWhithoutTask();
```
Comment (cjakeman) [inline Source/Menu/DownloadContentForm.cs:307]:
```suggestion
            mainForm.LoadFolderListWithoutTask();
```

Comment (cjakeman) [reply Source/Menu/DownloadContentForm.cs:307]: Just a typo

File: Source/Documentation/Manual/start.rst [hunk 4]
```diff
@@ -16,7 +16,24 @@ installation profile.
...
+
+.. image:: images/download-content-form.png
+
+Select the route to be installed. This list will be enlarged by the OR team when suitable routes become available.
+However it is still possible to install routes manually as described in the next chapter.
+
+* ``Install path`` and ``Browse``: Fill in or Browse to the path where the route is to be installed. Each route will be installed in its own subdirectory.
+* ``Info``: Gives more information of the selected route. The default web browser will be started with the information page.
+* ``Install``: Installs the selected route. A popup mentions the directory where the route is to be installed before starting the actual install. Installation profile gets added also.
+* ``Start``: Some routes have default start parameters and can be started via this button with these parameters.
```
Comment (cjakeman) [inline Source/Documentation/Manual/start.rst:34]:
```suggestion
```

Comment (cjakeman) [reply Source/Documentation/Manual/start.rst:34]: This option has been removed.

**Comment by cjakeman** [comment]
Hi Siebren,
> 
> I'll post some more at the weekend to see what your thoughts are. James and I are happy to do some of the work if necessary to get this released.

I'm very pleased that you have provided all this new functionality. It's going to make "Getting Started" so much easier for a newcomer.

Our main concern for this PR has been where the data for the installed routes is kept. I see that you have moved it from AppData\Roaming\OpenRails\Settings into the Registry, so that's all sorted. Thanks.

I've now submitted a review with a few minor points.

Once this PR is merged, I would like to discuss another PR for changes prior to releasing as v1.6:

- We would like to see the Options > Content tab merged into your new form - something like this:

image

- Any selection of a route should change the selection in the main form too - a behaviour that has been overlooked for years.

- The new form seems to be compressed vertically. James has offered to tidy it up to match the main form better.

- Perhaps we could discuss the choice of information shown in DownloadContent. For example, the author or the date might be more useful than the URL.

- Perhaps we could find an alternative to the Info button to present the data without requiring a browser.

If you would be willing to undertake another PR along these lines, that would be terrific. If not, then perhaps I can help out.

Best wishes,

Chris

**Review by cjakeman — CHANGES_REQUESTED** [review]
Hi Siebren,

Please remove the redundant statement about Start from the Manual.

The DownloadContent form has a resize control which does nothing. Please remove that also.

Then I'll approve and merge it.

Thanks,

Chris

File: Source/Menu/Options.cs [hunk 5]
```diff
@@ -728,11 +744,11 @@ private void textBoxContentName_TextChanged(object sender, EventArgs e)
             var current = bindingSourceContent.Current as ContentFolder;
             if (current != null && current.Name != textBoxContentName.Text)
             {
-                if (current.Path.ToLower().Contains(Application.StartupPath.ToLower()))
+                if (current.Path.ToLower().Contains(ApplicationInfo.ProcessDirectory.ToLower()))
```
[the remark this answers is not in the document]
Comment (sweiland-openrails) [reply Source/Menu/Options.cs:747]: I remember some merge conflict during the build of the unstable version. Don't knw precisely what went wrong there. If you want me find out please let me know.

File: Source/Menu/DownloadContentForm.cs (hunk 3 above)
Comment (sweiland-openrails) [reply Source/Menu/DownloadContentForm.cs]: Solved in https://github.com/openrails/openrails/pull/903/commits/7b5496d13bf0053a5a3c0b2700bb236424aeb3da

**Review by cjakeman — APPROVED** [review]
Hi Siebren,

Thanks for resolving those issues,

Approved and merged,

Chris

**Pull request merged**